himbaechel/uarch/gowin/pack.cc

```diff
@@ -640,6 +654,334 @@ struct GowinPacker
+
+            // input reg in IO
+            CellInfo *iologic_i = nullptr;
+            if ((ci.type == id_IBUF && ctx->settings.count(id_IREG_IN_IOB)) ||
+                (ci.type == id_IOBUF && ctx->settings.count(id_IOREG_IN_IOB))) {
+
+                if (ci.getPort(id_O) == nullptr) {
+                    continue;
+                }
+                // OBUF O -> D FF
+                CellInfo *ff = net_only_drives(ctx, ci.ports.at(id_O).net, is_ff, id_D);
+                if (ff == nullptr) {
+                    if (ci.attrs.count(id_IOBFF)) {
+                        log_warning("Port O of %s is not connected to FF.\n", ctx->nameOf(&ci));
+                    }
+                    continue;
+                }
+                if (ci.ports.at(id_O).net->users.entries() != 1) {
+                    if (ci.attrs.count(id_IOBFF)) {
+                        log_warning("Port O of %s is the driver of %s multi-sink network.\n", ctx->nameOf(&ci),
+                                    ctx->nameOf(ci.ports.at(id_O).net));
+                    }
+                    continue;
+                }
+                BelId l_bel = get_iologici_bel(&ci);
+                if (l_bel == BelId()) {
+                    continue;
+                }
+                if (ctx->debug) {
+                    log_info(" trying %s ff as Input Register of %s IO\n", ctx->nameOf(ff), ctx->nameOf(&ci));
+                }
+
+                clk_net = ff->getPort(id_CLK);
+                ce_net = ff->getPort(id_CE);
+                for (IdString port : {id_SET, id_RESET, id_PRESET, id_CLEAR}) {
+                    lsr_net = ff->getPort(port);
+                    if (lsr_net != nullptr) {
+                        break;
+                    }
+                }
+                reg_type = ff->type;
+
+                // create IOLOGIC cell for flipflop
+                IdString iologic_name = gwu.create_aux_name(ci.name, 0, "_iobff$");
+                auto iologic_cell = gwu.create_cell(iologic_name, id_IOLOGICI_EMPTY);
+                new_cells.push_back(std::move(iologic_cell));
+                iologic_i = new_cells.back().get();
+
+                // move ports
+                for (auto &port : ff->ports) {
+                    IdString port_name = port.first;
+                    ff->movePortTo(port_name, iologic_i, port_name != id_Q ? port_name : id_Q4);
+                }
+                if (ctx->verbose) {
+                    log_info("  place FF %s into IBUF %s, make iologic_i %s\n", ctx->nameOf(ff), ctx->nameOf(&ci),
+                             ctx->nameOf(iologic_i));
+                }
+                iologic_i->setAttr(id_HAS_REG, 1);
+                iologic_i->setAttr(id_IREG_TYPE, ff->type.str(ctx));
+                cells_to_remove.push_back(ff->name);
+            }
+
+            // output reg in IO
+            CellInfo *iologic_o = nullptr;
+            if ((ci.type == id_OBUF && ctx->settings.count(id_OREG_IN_IOB)) ||
```

Comment:
I think that FFs with the IOBFF attribute set should be packed into IO buffers regardless of the nextpnr configuration, this is perhaps the most useful approach when you only have a few flipflops that you are certain you want placed in them

Reply:
Done